pallets/oracle-data-collection/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+//! Oracle pallet to collect and aggregate oracle values.
+//!
+//! The collection admin configures the collection allowing a list of the oracle
+//! keys and from which feeders the values are collected.
+//!
+//! Later, the updating a collection will collect all values based on the admin
+//! configuration of the collection. The resulting collection is optimized to
+//! iterate though all values in jus one read.
+//!
+//! This pallet is not fed with external values, you need to configure a
+//! `ValueProvider` to provide it with oracle values.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cfg_traits::{data::DataRegistry, PreConditions, ValueProvider};
+	use frame_support::{pallet_prelude::*, storage::bounded_btree_map::BoundedBTreeMap};
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::{EnsureAddAssign, EnsureSubAssign, Zero};
+	use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
+
+	use crate::{
+		types::{CachedCollection, KeyInfo},
+		util,
+	};
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Identify an oracle value
+		type CollectionId: Parameter + Member + Copy + MaxEncodedLen;
+
+		/// Identify an oracle value
+		type OracleKey: Parameter + Member + Copy + MaxEncodedLen + Ord;
+
+		/// Represent an oracle value
+		type OracleValue: Parameter + Member + Copy + MaxEncodedLen + Ord;
+
+		/// Represent the a time moment
+		type Moment: Parameter + Member + Copy + MaxEncodedLen + Ord;
+
+		/// A way to obtain oracle values from feeders
+		type OracleProvider: ValueProvider<
+			(Self::AccountId, Self::CollectionId),
```

Comment:
Why this? THis will enforce that oracle providers use the `CollectionId` while feeding the oracles right?

Reply:
This is because of the OracleConverterBridge. From the data collection perspective, we need to query with the `CollectionId` in order to be able to make the conversion. But from oracle perspective, there is no `CollectionId`

Reply:
nice! THen we aggregate in the runtime?